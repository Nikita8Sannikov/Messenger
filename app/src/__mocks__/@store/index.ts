import useAuthStore from "./auth";
import useGlobalStore from "./global";
import useImagesCarouselStore from "./images-carousel";
import usePhotosStore from "./photos";
import useProfileStore from "./profile";
import useUIStore from "./ui";
import useUserStore from "./user";

const resetAllStores = jest.fn(() => {
	useUserStore.getState().reset();
	useGlobalStore.getState().reset();
	useUIStore.getState().reset();
	useImagesCarouselStore.getState().reset();
	useAuthStore.getState().reset();
	useProfileStore.getState().reset();
	usePhotosStore.getState().reset();
});

export default resetAllStores; 