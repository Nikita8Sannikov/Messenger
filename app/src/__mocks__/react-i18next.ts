export const useTranslation = () => ({
    t: (key: string) => {
        const dictionary: Record<string, string> = {
            "menu.profile": " My profile",
            "menu.messanger": "Messenger",
            "menu.friends": "Friends",
            "menu.to-developers": "To developers",
            "header.settings": "Settings",
            "header.help": "Help",
            "header.logout": "Log Out"
        };
        return dictionary[key];
    },
    i18n: {
        changeLanguage: jest.fn(),
        use: jest.fn().mockReturnThis(), // Мок для вызова use()
        init: jest.fn(), // Мок для вызова init()
    },
})

export const initReactI18next = {
    type: "3rdParty",
    init: jest.fn(),
}