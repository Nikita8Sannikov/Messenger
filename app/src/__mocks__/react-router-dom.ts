import { createElement, type ReactNode } from "react";
import * as ReactRouterDom from "react-router-dom";

const mockNavigate = jest.fn();
const mockLocation = {
	pathname: "/",
	search: "",
	hash: "",
	state: null,
	key: "default",
};

const mockParams = {};

const useNavigate = () => mockNavigate;
const useLocation = () => mockLocation;
const useParams = () => mockParams;

const matchPath = (pattern: string, pathname: string) => {
	if (pattern === pathname) {
		return {
			params: {},
			pathname,
			pattern,
			pathnameBase: pathname,
		};
	}
	return null;
};

const Routes = ({ children }: { children: ReactNode }) => children;
const Route = ({ children }: { children: ReactNode }) => children;
const Navigate = ({ to }: { to: string }) => createElement("div", { "data-testid": "navigate", "data-to": to });

export { useNavigate, mockNavigate, useLocation, mockLocation, useParams, mockParams, matchPath, Routes, Route, Navigate };
export default {
	...ReactRouterDom,
	useNavigate,
	useLocation,
	useParams,
	matchPath,
	Routes,
	Route,
	Navigate,
}; 