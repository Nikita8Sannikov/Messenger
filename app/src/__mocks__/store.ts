// import { configureStore } from "@reduxjs/toolkit";

// export const mockStore = configureStore({ reducer: () => ({}) });
