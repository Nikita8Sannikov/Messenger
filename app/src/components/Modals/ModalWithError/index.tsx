import React from "react";
import { useNavigate } from "react-router-dom";
import Alert from "@mui/material/Alert";
import Modal from "@mui/material/Modal";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import Button from "@mui/material/Button";
import RefreshIcon from "@mui/icons-material/Refresh";

import { useAppDispatch, useAppSelector } from "../../../hooks/useGlobalState";
import useMainClient from "../../../hooks/useMainClient";
import { selectErrorState, setError } from "../../../state/error/slice";
import { MAIL_FEEDBACK } from "../../../utils/constants";
import SnackBarComponent from "../../Common/Snackbar";

import "./modal-with-error.scss";

const modalTitle = "modal-error-title";
const modalDescription = "modal-error-description";

export default function ModalWithError() {
    const [open, setOpen] = React.useState(false);
    const [visible, setVisible] = React.useState(false);

    const mainClient = useMainClient();

    const { error } = useAppSelector(selectErrorState);
    const dispatch = useAppDispatch();
    const navigate = useNavigate();

    React.useEffect(() => {
        setOpen(Boolean(error));
    }, [error]);

    // Копирование текста в буфер обмена
    const onCopy = () => {
        if (navigator.clipboard && error) {
            navigator.clipboard.writeText(error)
                .then(() => {
                    setVisible(true);
                })
                .catch((error: Error) => {
                    mainClient.catchErrors("Ошибка при копировании текста в буфер обмена: " + error);
                });
        }
    };

    // Закрытие модального окна
    const onClose = (_: Object, reason: string) => {
        if (reason !== "backdropClick") {
            setOpen(false);
            dispatch(setError(null));
        }
    };

    // Обновление страницы
    const onReload = () => {
        navigate(0);
    };

    return <>
        <SnackBarComponent anchor={{ vertical: "top", horizontal: "center" }} open={visible} handleClose={() => setVisible(false)}>
            <Alert className="alert-error-container" onClose={() => setVisible(false)} severity="success">
                Текст успешно скопирован в буфер обмена!
            </Alert>
        </SnackBarComponent>

        <Modal 
            open={open} 
            onClose={onClose} 
            aria-labelledby={modalTitle} 
            aria-describedby={modalDescription}
            disableEscapeKeyDown
        >
            <Box className="modal-error-container">
                <Typography variant="h6" component="h2">
                    Упс! Возникла ошибка в работе сервера
                </Typography>

                <Typography className="modal-error-container__text">
                    Пожалуйста, скопируйте текст ошибки и отправьте её на почту разработчикам: {MAIL_FEEDBACK}
                </Typography>

                <div className="modal-error-container__error" onClick={onCopy}>
                    {error}
                </div>

                <div className="modal-error-container__button">
                    <Button variant="outlined" startIcon={<RefreshIcon />} onClick={onReload}>Обновить страницу</Button>
                </div>
            </Box>
        </Modal>
    </>
};