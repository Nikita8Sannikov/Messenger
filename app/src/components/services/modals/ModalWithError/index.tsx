import { useState, useEffect, useContext, memo } from "react";
import { useNavigate } from "react-router-dom";
import { useTranslation } from "react-i18next";
import Alert from "@mui/material/Alert";
import Modal from "@mui/material/Modal";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import Button from "@mui/material/Button";
import RefreshIcon from "@mui/icons-material/Refresh";

import SnackBarComponent from "../../../ui/Snackbar";
import { MainClientContext } from "../../../main/Main";
import { useAppDispatch, useAppSelector } from "../../../../hooks/useGlobalState";
import { selectErrorState, setError } from "../../../../store/error/slice";
import { MAIL_FEEDBACK } from "../../../../utils/constants";

import "./modal-with-error.scss";

const modalTitle = "modal-error-title";
const modalDescription = "modal-error-description";

export default memo(function ModalWithError() {
    const [open, setOpen] = useState(false);
    const [visible, setVisible] = useState(false);

    const mainClient = useContext(MainClientContext);

    const { t } = useTranslation();
    const { error } = useAppSelector(selectErrorState);
    const dispatch = useAppDispatch();
    const navigate = useNavigate();

    useEffect(() => {
        setOpen(Boolean(error));
    }, [error]);

    // Копирование текста в буфер обмена
    const onCopy = () => {
        if (navigator.clipboard && error) {
            navigator.clipboard.writeText(error)
                .then(() => {
                    setVisible(true);
                })
                .catch((error: Error) => {
                    mainClient.catchErrors(t("modals.error.copy") + error);
                });
        }
    };

    // Закрытие модального окна
    const onClose = (_: Object, reason: string) => {
        if (reason !== "backdropClick") {
            setOpen(false);
            dispatch(setError(null));
        }
    };

    // Обновление страницы
    const onReload = () => {
        navigate(0);
    };

    return <>
        <SnackBarComponent anchor={{ vertical: "top", horizontal: "center" }} open={visible} handleClose={() => setVisible(false)}>
            <Alert className="alert-error-container" onClose={() => setVisible(false)} severity="success">
                { t("modals.copy_successfull") }
            </Alert>
        </SnackBarComponent>

        <Modal 
            open={open} 
            onClose={onClose} 
            aria-labelledby={modalTitle} 
            aria-describedby={modalDescription}
            disableEscapeKeyDown
        >
            <Box className="modal-error-container">
                <Typography variant="h6" component="h2">
                    { t("modals.error.server") }
                </Typography>

                <Typography className="modal-error-container__text">
                    { t("modals.copy_the_message_and_send_to_email", { email: MAIL_FEEDBACK }) }
                </Typography>

                <div className="modal-error-container__error" onClick={onCopy}>
                    {error}
                </div>

                <div className="modal-error-container__button">
                    <Button variant="outlined" startIcon={<RefreshIcon />} onClick={onReload}>{ t("modals.reset_page") }</Button>
                </div>
            </Box>
        </Modal>
    </>
});