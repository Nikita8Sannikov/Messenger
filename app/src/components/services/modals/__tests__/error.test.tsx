import { act, fireEvent, render, screen } from "@testing-library/react";

import ModalWithError from "@components/services/modals/error";
import mockUseMainClient from "../../../../__mocks__/@hooks/useMainClient";
import mockUIStore from "../../../../__mocks__/@store/ui";
import { mockNavigate } from "../../../../__mocks__/react-router-dom";

Object.assign(navigator, {
	clipboard: {
		writeText: jest.fn(),
	},
});

describe("ModalWithError", () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it("should not render when error is null", () => {
		const store = mockUIStore.getState();
		store.error = null;

		const { container } = render(<ModalWithError />);
		expect(container).toBeEmptyDOMElement();
	});

	it("should render with error message", () => {
		const store = mockUIStore.getState();
		store.error = "Test error message";

		const { container } = render(<ModalWithError />);
        
		expect(screen.getByText("Oops! An error occurred in the server operation.")).toBeInTheDocument();
		expect(screen.getByText("Test error message")).toBeInTheDocument();
		expect(screen.getByText("Download the log file")).toBeInTheDocument();
		expect(screen.getByText("Reset page")).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it("should copy error message to clipboard when clicked", async () => {
		const store = mockUIStore.getState();
		store.error = "Test error message";
		(navigator.clipboard.writeText as jest.Mock).mockResolvedValue(undefined);

		render(<ModalWithError />);
        
		const errorText = screen.getByText("Test error message");
		await act(async () => {
			await fireEvent.click(errorText);
		});

		expect(navigator.clipboard.writeText).toHaveBeenCalledWith("Test error message");
		await screen.findByText("The text has been successfully copied to the clipboard!");
	});

	it("should show error when clipboard write fails", async () => {
		const store = mockUIStore.getState();
		store.error = "Test error message";
		const clipboardError = new Error("Clipboard error");
		(navigator.clipboard.writeText as jest.Mock).mockRejectedValue(clipboardError);

		render(<ModalWithError />);
        
		const errorText = screen.getByText("Test error message");
		await act(async () => {
			await fireEvent.click(errorText);
		});

		expect(store.setError).toHaveBeenCalledWith("Error when copying text to clipboard: Error: Clipboard error");
	});

	it("should download log file when download button is clicked", () => {
		const store = mockUIStore.getState();
		store.error = "Test error message";

		render(<ModalWithError />);
        
		const downloadButton = screen.getByText("Download the log file");
		fireEvent.click(downloadButton);

		expect(mockUseMainClient().downloadLogFile).toHaveBeenCalled();
	});

	it("should reload page when reset button is clicked", () => {
		const store = mockUIStore.getState();
		store.error = "Test error message";

		render(<ModalWithError />);
        
		const resetButton = screen.getByText("Reset page");
		fireEvent.click(resetButton);

		expect(mockNavigate).toHaveBeenCalledWith(0);
	});

	it("should close modal when close button is clicked", () => {
		const store = mockUIStore.getState();
		store.error = "Test error message";

		render(<ModalWithError />);
        
		const closeButton = screen.getByTestId("close-icon");
		fireEvent.click(closeButton);

		expect(store.setError).toHaveBeenCalledWith(null);
	});

	it("should render with email in copy message", () => {
		const store = mockUIStore.getState();
		store.error = "Test error message";

		render(<ModalWithError />);
        
		expect(screen.getByText("Please copy the error text and email it to the developers: mocked@mail.com.")).toBeInTheDocument();
	});

	it("should handle empty support email", () => {
		const store = mockUIStore.getState();
		store.error = "Test error message";
		const originalEmail = process.env.REACT_APP_SUPPORT_EMAIL;
		process.env.REACT_APP_SUPPORT_EMAIL = "";

		render(<ModalWithError />);
        
		expect(screen.getByText("Please copy the error text and email it to the developers: mocked@mail.com.")).toBeInTheDocument();
        
		process.env.REACT_APP_SUPPORT_EMAIL = originalEmail;
	});
});
