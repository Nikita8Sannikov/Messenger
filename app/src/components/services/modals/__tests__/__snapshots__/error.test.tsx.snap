// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalWithError should render with error message 1`] = `<div />`;
